include/chainbase/shared_cow_vector.hpp

```diff
@@ -0,0 +1,266 @@
+#pragma once
+#include <boost/interprocess/offset_ptr.hpp>
+#include <boost/interprocess/allocators/allocator.hpp>
+
+#include <cstddef>
+#include <cstring>
+#include <algorithm>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include <chainbase/pinnable_mapped_file.hpp>
+
+namespace chainbase {
+   namespace bip = boost::interprocess;
+
+   template<typename T>
+   class shared_cow_vector {
+      struct impl {
+         uint32_t reference_count;
+         uint32_t size;
+         T data[0];
+      };
+
+   public:
+      using allocator_type = bip::allocator<char, segment_manager>;
+      using iterator       = const T*;
+      using const_iterator = const T*;
+
+      explicit shared_cow_vector() = default;
+
+      template<typename Iter>
+      explicit shared_cow_vector(Iter begin, Iter end) {
+         std::size_t size = std::distance(begin, end);
+         _alloc<false>(&*begin, size, size);
+      }
+
+      explicit shared_cow_vector(const T* ptr, std::size_t size) {
+         _alloc<false>(ptr, size, size);
+      }
+
+      shared_cow_vector(const shared_cow_vector& o) {
+         if (get_allocator(this) == o.get_allocator()) {
+            _data = o._data;
+            if (_data != nullptr)
+               ++_data->reference_count;
+         } else {
+            if (o._data)
+               std::construct_at(this, o.data(),  o.size()); // call other constructor
+         }
+      }
+
+      shared_cow_vector(shared_cow_vector&& o) noexcept {
+         if (get_allocator() == o.get_allocator()) {
+            _data = o._data;
+            o._data = nullptr;
+         } else {
+            if (o._data)
+               std::construct_at(this, o.data(),  o.size());
+         }
+      }
+
+      template<class I, std::enable_if_t<std::is_constructible_v<T, I>, int> = 0 >
+      explicit shared_cow_vector(std::initializer_list<I> init) {
+         clear_and_construct(init.size(), 0, [&](T* dest, std::size_t idx) {
+            std::construct_at(dest, std::data(init)[idx]);
+         });
+      }
+
+      template<class I, std::enable_if_t<std::is_constructible_v<T, I>, int> = 0 >
+      explicit shared_cow_vector(const std::vector<I>& v) {
+         clear_and_construct(v.size(), 0, [&](T* dest, std::size_t idx) {
+            std::construct_at(dest, v[idx]);
+         });
+      }
+
+      template<class I, std::enable_if_t<std::is_constructible_v<T, I>, int> = 0 >
```

Comment:
Shouldn't this be `std::is_move_constructible_v`

Reply:
`std::is_move_constructible_v` has a different use, it takes only one template parameter and checks if this type is move_constructible itself. I think the code as it is is correct.